fix: Helm - Propagation of maxBackgroundReports

mihaigalos opened the pull request:
## Explanation
https://github.com/kyverno/kyverno/pull/11329 introduced `maxBackgroundReports` to limit the number of EphemeralReports CRs. This is known to cause issues in ArgoCD - [ref](https://github.com/argoproj/argo-cd/issues/20977).

The original PR however did not update the Helm chart for propagating the field to the controller.
<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->

## Related issue

<!--
Please link the GitHub issue this pull request resolves in the format of `Closes #1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## Documentation (required for features)
No new features, just wiring of Helm values to the controller.

## What type of PR is this

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->
/kind bug

## Proposed Changes
Just propagate a Helm flag to the consuming controllers.
<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

<details><summary> generated deployment.yaml</summary>
<p>

```yaml
# Source: kyverno/templates/background-controller/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: kyverno-background-controller
  namespace: default
  labels:
    app.kubernetes.io/component: background-controller
    app.kubernetes.io/instance: release-name
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: release-name-kyverno
    app.kubernetes.io/version: v0.0.0
    helm.sh/chart: kyverno-v0.0.0
spec:
  replicas: 
  revisionHistoryLimit: 10
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 40%
    type: RollingUpdate
  selector:
    matchLabels:
      app.kubernetes.io/component: background-controller
      app.kubernetes.io/instance: release-name
      app.kubernetes.io/part-of: release-name-kyverno
  template:
    metadata:
      labels:
        app.kubernetes.io/component: background-controller
        app.kubernetes.io/instance: release-name
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/part-of: release-name-kyverno
        app.kubernetes.io/version: v0.0.0
        helm.sh/chart: kyverno-v0.0.0
    spec:
      dnsPolicy: ClusterFirst
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchExpressions:
                - key: app.kubernetes.io/component
                  operator: In
                  values:
                  - background-controller
              topologyKey: kubernetes.io/hostname
            weight: 1
      serviceAccountName: kyverno-background-controller
      containers:
        - name: controller
          image: "reg.kyverno.io/kyverno/background-controller:latest"
          imagePullPolicy: IfNotPresent
          ports:
          - containerPort: 9443
            name: https
            protocol: TCP
          - containerPort: 8000
            name: metrics
            protocol: TCP
          
          args:
            - --disableMetrics=false
            - --otelConfig=prometheus
            - --metricsPort=8000
            - --resyncPeriod=15m
            - --enableConfigMapCaching=true
            - --enableDeferredLoading=true
            - --maxAPICallResponseLength=2000000
            - --maxBackgroundReports=10000
            - --loggingFormat=text
            - --v=2
            - --omitEvents=PolicyApplied,PolicySkipped
            - --enablePolicyException=false
            - --enableReporting=validate,mutate,mutateExisting,imageVerify,generate
            
          env:
          - name: KYVERNO_SERVICEACCOUNT_NAME
            value: kyverno-background-controller
          - name: KYVERNO_DEPLOYMENT
            value: kyverno-background-controller
          - name: INIT_CONFIG
            value: release-name-kyverno
          - name: METRICS_CONFIG
            value: release-name-kyverno-metrics
          - name: KYVERNO_POD_NAME
            valueFrom:
              fieldRef:
                fieldPath: metadata.name
          - name: KYVERNO_NAMESPACE
            valueFrom:
              fieldRef:
                fieldPath: metadata.namespace
          resources:
            limits:
              memory: 128Mi
            requests:
              cpu: 100m
              memory: 64Mi
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop:
              - ALL
            privileged: false
            readOnlyRootFilesystem: true
            runAsNonRoot: true
            seccompProfile:
              type: RuntimeDefault
```

</p>
</details> 

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [ ] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [ ] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

realshuting commented:
Thanks @mihaigalos for the PR. Can you add this change to [the Helm changelog](https://github.com/rsafonseca/kyverno/blob/0548d09c21c97d4c16f93ca5a40878534b2d4f4e/charts/kyverno/Chart.yaml#L35) (let's remove old ones as well)?

mihaigalos commented:
Hi @realshuting. Done.

mihaigalos commented:
Had a look at the code, seems there is no cleanup job for this, only a breaker to stop further additional reports from being produced.
Any plans for the cleanup?

realshuting commented:
> Had a look at the code, seems there is no cleanup job for this, only a breaker to stop further additional reports from being produced. Any plans for the cleanup?

Nope, we removed the cleanup job in 1.12 and added the circuit breaker in 1.13.

realshuting reviewed (dismissed):
/lgtm

mihaigalos commented:
> > Had a look at the code, seems there is no cleanup job for this, only a breaker to stop further additional reports from being produced. Any plans for the cleanup?
> 
> Nope, we removed the cleanup job in 1.12 and added the circuit breaker in 1.13.

ok. So existing EphemeralReports peak at the threshold and the queue is appended to at the beginning with new CRs.
This, however, doesn't guard against 1 Namespace having a lot of ERs and others having a low number.
It wouldn't help alleviate the ArgoCD symptoms with it freezing because of high amount of ERs in an app.

realshuting commented:
> > > Had a look at the code, seems there is no cleanup job for this, only a breaker to stop further additional reports from being produced. Any plans for the cleanup?
> > 
> > 
> > Nope, we removed the cleanup job in 1.12 and added the circuit breaker in 1.13.
> 
> ok. So existing EphemeralReports peak at the threshold and the queue is appended to at the beginning with new CRs. This, however, doesn't guard against 1 Namespace having a lot of ERs and others having a low number. It wouldn't help alleviate the ArgoCD symptoms with it freezing because of high amount of ERs in an app.


>It wouldn't help alleviate the ArgoCD symptoms with it freezing because of high amount of ERs in an app.

Can you elaborate on the issue with ArgoCD? We can create an issue and continue the discussion there.

mihaigalos commented:
> > > > Had a look at the code, seems there is no cleanup job for this, only a breaker to stop further additional reports from being produced. Any plans for the cleanup?
> > > 
> > > 
> > > Nope, we removed the cleanup job in 1.12 and added the circuit breaker in 1.13.
> > 
> > 
> > ok. So existing EphemeralReports peak at the threshold and the queue is appended to at the beginning with new CRs. This, however, doesn't guard against 1 Namespace having a lot of ERs and others having a low number. It wouldn't help alleviate the ArgoCD symptoms with it freezing because of high amount of ERs in an app.
> 
> > It wouldn't help alleviate the ArgoCD symptoms with it freezing because of high amount of ERs in an app.
> 
> Can you elaborate on the issue with ArgoCD? We can create an issue and continue the discussion there.

Nevermind, seems this is unrelated.

realshuting commented:
Controllers failed to build, and code-gen needs updates https://github.com/kyverno/kyverno/actions/runs/13384535240/job/37380322747?pr=12186.

realshuting commented:
Looks like the controller fails to build.

mihaigalos commented:
Is this related to my change? Looks like some leftovers when bumping CRDs.

mihaigalos commented:
seems only the backgroundController has this `maxBackgroundReports` field - logs from admissionController `1.13.4`:
```
flag provided but not defined: -maxBackgroundReports
Usage of /ko-app/kyverno:
...
```
